pkg/utils/ldap/reconnect.go

```diff
@@ -0,0 +1,341 @@
+				l, err = c.ldapConnect(config)
+			} else {
+				c.logger.Debug().Msg("already reconnected")
+			}
+		case c.conn <- ldapConnection{l, err}:
+		}
+	}
+}
+
+func (c *ConnWithReconnect) ldapConnect(config Config) (*ldap.Conn, error) {
+	c.logger.Debug().Msgf("Connecting to %s", config.URI)
+
+	var err error
+	var l *ldap.Conn
+	if config.TLSConfig != nil {
+		l, err = ldap.DialURL(config.URI, ldap.DialWithTLSConfig(config.TLSConfig))
+	} else {
+		l, err = ldap.DialURL(config.URI)
+	}
+
+	if err != nil {
+		c.logger.Debug().Err(err).Msg("could not get ldap Connection")
+	} else {
+		c.logger.Debug().Msg("LDAP Connected")
+		if config.BindDN != "" {
+			c.logger.Debug().Msgf("Binding as %s", config.BindDN)
+			err = l.Bind(config.BindDN, config.BindPassword)
+			if err != nil {
+				c.logger.Debug().Err(err).Msg("Bind failed")
+				l.Close()
+				return nil, err
+			}
+
+		}
+	}
+
+	return l, err
+}
+
+func (c *ConnWithReconnect) reconnect(resetConn *ldap.Conn) (*ldap.Conn, error) {
+	c.logger.Debug().Msg("LDAP connection reset")
+	c.reset <- resetConn
+	c.logger.Debug().Msg("Waiting for new connection")
+	result := <-c.conn
+	return result.Conn, result.Error
+}
+
+// Remaining methods to fulfill ldap.Client interface
+
+// Start implements the ldap.Client interface
+func (c *ConnWithReconnect) Start() {}
+
+// StartTLS implements the ldap.Client interface
+func (c *ConnWithReconnect) StartTLS(*tls.Config) error {
+	return ldap.NewError(ldap.LDAPResultNotSupported, fmt.Errorf("not implemented"))
+}
+
+// Close implements the ldap.Client interface
+func (c *ConnWithReconnect) Close() {}
+
+// IsClosing implements the ldap.Client interface
+func (c *ConnWithReconnect) IsClosing() bool {
+	return false
+}
+
+// SetTimeout implements the ldap.Client interface
+func (c *ConnWithReconnect) SetTimeout(time.Duration) {}
+
+// Bind implements the ldap.Client interface
+func (c *ConnWithReconnect) Bind(username, password string) error {
+	return ldap.NewError(ldap.LDAPResultNotSupported, fmt.Errorf("not implemented"))
+}
+
+// UnauthenticatedBind implements the ldap.Client interface
+func (c *ConnWithReconnect) UnauthenticatedBind(username string) error {
+	return ldap.NewError(ldap.LDAPResultNotSupported, fmt.Errorf("not implemented"))
+}
+
+// SimpleBind implements the ldap.Client interface
+func (c *ConnWithReconnect) SimpleBind(*ldap.SimpleBindRequest) (*ldap.SimpleBindResult, error) {
+	return nil, ldap.NewError(ldap.LDAPResultNotSupported, fmt.Errorf("not implemented"))
+}
+
+// ExternalBind implements the ldap.Client interface
+func (c *ConnWithReconnect) ExternalBind() error {
+	return ldap.NewError(ldap.LDAPResultNotSupported, fmt.Errorf("not implemented"))
+}
+
+// ModifyWithResult implements the ldap.Client interface
+func (c *ConnWithReconnect) ModifyWithResult(m *ldap.ModifyRequest) (*ldap.ModifyResult, error) {
+	conn, err := c.getConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	return conn.ModifyWithResult(m)
```

Comment:
Curious: Would this work for all Methods? So let's say I want to Implement `SetTimeout` Method. Would I just do:
```golang
conn, err := c.getConnection()
if err != nil {
	return
}

return conn.SetTimeout(t)
```

Reply:
No. This one slipped through from an earlier experiment. It's lacking the retry. But for now we don't even need the `ModifyWithResult` call.